x/expiration/keeper/keeper.go

```diff
@@ -0,0 +1,565 @@
+		return nil, types.ErrInvalidMessage.Wrapf("failed to unpack msg: %v", err)
+	}
+	if err := msg.ValidateBasic(); err != nil {
+		return nil, types.ErrInvalidMessage.Wrapf("failed ValidateBasic: %v", err)
+	}
+
+	// anyone can delete an expired expiration
+	if ctx.BlockTime().After(expiration.Time) {
+		return expiration, nil
+	}
+
+	// validate signers and owner
+	if err := k.validateSigners(ctx, expiration.Owner, signers, msgTypeURL); err != nil {
+		return expiration, types.ErrInvalidSigners.Wrap(err.Error())
+	}
+
+	return expiration, nil
+}
+
+// ResolveDepositor resolves the account address where the expiration message deposit will be refunded.
+func (k Keeper) ResolveDepositor(
+	ctx sdk.Context,
+	expiration types.Expiration,
+	msg *types.MsgInvokeExpirationRequest,
+) (sdk.AccAddress, error) {
+	// check for owner in list first
+	for _, signer := range msg.Signers {
+		if signer == expiration.Owner {
+			addr, err := sdk.AccAddressFromBech32(expiration.Owner)
+			if err != nil {
+				return nil, types.ErrResolveDepositor.Wrap(
+					types.ErrInvalidSigners.Wrapf("[%s]: %v", signer, err).Error())
+			}
+			return addr, nil
+		}
+	}
+	// fall back to first signer if after expiration
+	if ctx.BlockTime().After(expiration.Time) && len(msg.Signers) >= 1 {
+		signer := msg.Signers[0]
+		addr, err := sdk.AccAddressFromBech32(signer)
+		if err != nil {
+			return nil, types.ErrResolveDepositor.Wrap(
+				types.ErrInvalidSigners.Wrapf("[%s]: %v", signer, err).Error())
+		}
+		return addr, nil
+	}
+
+	// error if no qualifying depositors are found
+	return nil, types.ErrResolveDepositor.Wrapf("asset [%s]", msg.ModuleAssetId)
+}
+
+// Private method that validates expiration message signatures
+func (k Keeper) validateSigners(
+	ctx sdk.Context,
+	owner string,
+	signers []string,
+	msgTypeURL string,
+) error {
+	found := false
+	for _, signer := range signers {
+		if signer == owner {
+			found = true
+			break
+		}
+
+		// validate signer with authz
+		var err error
+		found, err = k.hasSignerWithAuthz(ctx, owner, signer, msgTypeURL)
+		if err != nil {
+			return err
+		}
+		if found {
+			break
+		}
+	}
+
+	if !found {
+		return fmt.Errorf("intended signers %s do not match given signer [%s]", signers, owner)
+	}
+
+	return nil
+}
+
+// Private method that checks if any of the signers have been
+// granted authorizations to perform actions on an expiration record
+func (k Keeper) hasSignerWithAuthz(
+	ctx sdk.Context,
+	owner string,
+	signer string,
+	msgTypeURL string,
+) (bool, error) {
+	granter, err := sdk.AccAddressFromBech32(owner)
+	if err != nil {
+		return false, fmt.Errorf("invalid owner: %w", err)
+	}
+	grantee, err := sdk.AccAddressFromBech32(signer)
+	if err != nil {
+		return false, fmt.Errorf("invalid signers: %w", err)
+	}
+
+	authorization, exp := k.authzKeeper.GetAuthorization(ctx, grantee, granter, msgTypeURL)
+	if authorization != nil {
+		resp, err := authorization.Accept(ctx, nil)
```

Comment:
The message being processed needs to be provided to this `Accept` call. We don't have any expectations about either the authorizations being used or the messages they're being applied to, so standard authz processing needs to be used.